File: yarl-java/src/test/java/hr/tjakopan/yarl/PolicyContextAndKeyTest.java
package hr.tjakopan.yarl;

import hr.tjakopan.yarl.retry.RetryPolicy;
import hr.tjakopan.yarl.test.helpers.PolicyUtils;
import hr.tjakopan.yarl.test.helpers.TestResult;
import kotlin.Unit;
import kotlin.jvm.functions.Function3;
import org.junit.Test;

import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

import static hr.tjakopan.yarl.Functions.fromConsumer3;
import static org.assertj.core.api.Assertions.assertThat;

public class PolicyContextAndKeyTest {
  //<editor-fold desc="configuration">
  @Test
  public void shouldBeAbleFluentlyToConfigureThePolicyKey() {
    final var policy = RetryPolicy.<Integer>builder()
      .policyKey(UUID.randomUUID().toString())
      .handleResult(0)
      .retry();

    assertThat(policy).isInstanceOf(Policy.class);
  }

  @Test
  public void policyKeyPropertyShouldBeTheFluentlyConfiguredPolicyKey() {
    final var key = "SomePolicyKey";
    final var policy = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .policyKey(key)
      .retry();

    assertThat(policy.getPolicyKey()).isEqualTo(key);
  }

  @Test
  public void policyKeyPropertyShouldBeNonNullOrEmptyIfNotExplicitlyConfigured() {
    final var policy = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .retry();

    assertThat(policy.getPolicyKey()).isNotNull();
    assertThat(policy.getPolicyKey()).isNotEmpty();
  }

  @Test
  public void policyKeyPropertyShouldStartWithPolicyTypeIfNotExplicitlyConfigured() {
    final var policy = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .retry();

    assertThat(policy.getPolicyKey()).startsWith("RetryPolicy");
  }

  @Test
  public void policyKeyPropertyShouldBeUniqueForDifferentInstancesIfNotExplicitlyConfigured() {
    final var policy1 = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .retry();
    final var policy2 = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .retry();

    assertThat(policy1.getPolicyKey()).isNotEqualTo(policy2.getPolicyKey());
  }

  @Test
  public void policyKeyPropertyShouldReturnConsistentValueForSamePolicyInstanceIfNotExplicitlyConfigured() {
    final var policy = RetryPolicy.<Integer>builder()
      .handleResult(0)
      .retry();

    final var keyRetrievedFirst = policy.getPolicyKey();
    final var keyRetrievedSecond = policy.getPolicyKey();

    assertThat(keyRetrievedSecond).isSameAs(keyRetrievedFirst);
  }

  @Test
  public void shouldPassPolicyKeyToExecutionContext() {
    final var policyKey = UUID.randomUUID().toString();
    final AtomicReference<String> policyKeySetOnExecutionContext = new AtomicReference<>();
    final Function3<DelegateResult<TestResult>, Integer, Context, Unit> onRetry =
      fromConsumer3(result -> (retryCount, context) -> policyKeySetOnExecutionContext.set(context.getPolicyKey()));
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .policyKey(policyKey)
      .retry(1, onRetry);

    PolicyUtils.raiseResults(policy, TestResult.FAULT, TestResult.GOOD);

    assertThat(policyKeySetOnExecutionContext.get()).isEqualTo(policyKey);
  }
  //</editor-fold>

  @Test
  public void shouldPassOperationKeyToExecutionContext() {
    final var operationKey = "SomeKey";
    final AtomicReference<String> operationKeySetOnContext = new AtomicReference<>();
    final Function3<DelegateResult<TestResult>, Integer, Context, Unit> onRetry =
      fromConsumer3(result -> (retryCount, context) -> operationKeySetOnContext.set(context.getOperationKey()));
    final var policy = RetryPolicy.<TestResult>builder()
      .handleResult(TestResult.FAULT)
      .retry(1, onRetry);

    final AtomicBoolean firstExecution = new AtomicBoolean(true);
    final var context = Context.builder()
      .operationKey(operationKey)
      .build();
    policy.execute(context, ctx -> {
      if (firstExecution.get()) {
        firstExecution.set(false);
        return TestResult.FAULT;
      }
      return TestResult.GOOD;
    });

    assertThat(operationKeySetOnContext.get()).isEqualTo(operationKey);
  }
}

File: yarl-core/src/test/kotlin/hr/tjakopan/yarl/PolicyContextAndKeyTest.kt
package hr.tjakopan.yarl

import hr.tjakopan.yarl.retry.retry
import hr.tjakopan.yarl.test.helpers.TestResult
import hr.tjakopan.yarl.test.helpers.raiseResults
import org.assertj.core.api.Assertions.assertThat
import java.util.*
import kotlin.test.Test

class PolicyContextAndKeyTest {
  //<editor-fold desc="configuration">
  @Test
  fun `should be able fluently to configure the policy key`() {
    val policy = Policy.retry<Int>()
      .policyKey(UUID.randomUUID().toString())
      .handleResult(0)
      .retry()

    assertThat(policy).isInstanceOf(Policy::class.java)
  }

  @Test
  fun `policy key property should be the fluently configured policy key`() {
    val key = "SomePolicyKey"
    val policy = Policy.retry<Int>()
      .handleResult(0)
      .policyKey(key)
      .retry()

    assertThat(policy.policyKey).isEqualTo(key)
  }

  @Suppress("UsePropertyAccessSyntax")
  @Test
  fun `policy key property should be non null or empty if not explicitly configured`() {
    val policy = Policy.retry<Int>()
      .handleResult(0)
      .retry()

    assertThat(policy.policyKey).isNotNull()
    assertThat(policy.policyKey).isNotEmpty()
  }

  @Test
  fun `policy key property should start with policy type if not explicitly configured`() {
    val policy = Policy.retry<Int>()
      .handleResult(0)
      .retry()

    assertThat(policy.policyKey).startsWith("RetryPolicy")
  }

  @Test
  fun `policy key property should be unique for different instances if not explicitly configured`() {
    val policy1 = Policy.retry<Int>()
      .handleResult(0)
      .retry()
    val policy2 = Policy.retry<Int>()
      .handleResult(0)
      .retry()

    assertThat(policy1.policyKey).isNotEqualTo(policy2.policyKey)
  }

  @Test
  fun `policy key property should return consistent value for same policy instance if not explicitly configured`() {
    val policy = Policy.retry<Int>()
      .handleResult(0)
      .retry()

    val keyRetrievedFirst = policy.policyKey
    val keyRetrievedSecond = policy.policyKey

    assertThat(keyRetrievedSecond).isSameAs(keyRetrievedFirst)
  }
  //</editor-fold>

  @Test
  fun `should pass policy key to execution context`() {
    val policyKey = UUID.randomUUID().toString()
    var policyKeySetOnExecutionContext: String? = null
    val onRetry: (DelegateResult<TestResult>, Int, Context) -> Unit =
      { _, _, context -> policyKeySetOnExecutionContext = context.policyKey }
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .policyKey(policyKey)
      .retry(1, onRetry)

    policy.raiseResults(TestResult.FAULT, TestResult.GOOD)

    assertThat(policyKeySetOnExecutionContext).isEqualTo(policyKey)
  }

  @Test
  fun `should pass operation key to execution context`() {
    val operationKey = "SomeKey"
    var operationKeySetOnContext: String? = null
    val onRetry: (DelegateResult<TestResult>, Int, Context) -> Unit =
      { _, _, context -> operationKeySetOnContext = context.operationKey }
    val policy = Policy.retry<TestResult>()
      .handleResult(TestResult.FAULT)
      .retry(1, onRetry)

    var firstExecution: Boolean = true
    policy.execute(Context(operationKey)) {
      if (firstExecution) {
        firstExecution = false
        return@execute TestResult.FAULT
      }
      return@execute TestResult.GOOD
    }

    assertThat(operationKeySetOnContext).isEqualTo(operationKey)
  }
}
